cms/extensions/toolbar.py

```diff
@@ -60,41 +70,61 @@ def get_page_extension_admin(self):
 
-        Return a list of tuples of the title extension and the url; the extension is None if no instance exists,
-        the url is None is no admin is registered for the extension.
+        Return a list of tuples of the page content extension and the url; the extension is None
+        if no instance exists, the url is None is no admin is registered for the extension.
         """
+        warnings.warn(
+            "get_title_extension_admin has been deprecated and replaced by get_page_content_extension_admin",
+            DeprecationWarning, stacklevel=2,
+        )
         page = self._get_page()
+
+        page_contents = page.pagecontent_set(manager="admin_manager").latest_content()\
+            .filter(language__in=get_language_list(page.node.site_id))
         urls = []
-        if language:
-            titles = page.get_content_obj(language),
-        else:
-            titles = page.pagecontent_set.filter(language__in=get_language_list(page.node.site_id))
-        # Titles
-        for title in titles:
-            try:
-                title_extension = self.model.objects.get(extended_object_id=title.pk)
-            except self.model.DoesNotExist:
-                title_extension = None
-            try:
-                model_name = self.model.__name__.lower()
-                if title_extension:
-                    admin_url = admin_reverse(
-                        '%s_%s_change' % (self.model._meta.app_label, model_name),
-                        args=(title_extension.pk,))
-                else:
-                    admin_url = "%s?extended_object=%s" % (
-                        admin_reverse('%s_%s_add' % (self.model._meta.app_label, model_name)),
-                        title.pk)
-            except NoReverseMatch:  # pragma: no cover
-                admin_url = None
+
+        for page_content in page_contents:
+            admin_url = self.get_page_content_extension_admin(page_content)
             if admin_url:
-                urls.append((title_extension, admin_url))
+                urls.append(admin_url)
         return urls
 
+        return self.get_page_content_extension_admin(language)
+
+    def get_page_content_extension_admin(self, page_content_obj=None):
+        """
+        Get the admin url for the page content extensions menu item, depending on whether a
+        :class:`~cms.extensions.models.PageContentExtension` instance exists for the
+        :class:`~cms.models.contentmodels.PageContent` displayed.
+
+        Return a tuple of the page content extension and the url; the extension is None
+        if no instance exists, the url is None is no admin is registered for the extension.
+        """
+        self._get_page()
+        page_content = page_content_obj or self.page_content
+        try:
+            pagecontent_extension = self.model.objects.get(extended_object_id=page_content.pk)
+        except self.model.DoesNotExist:
+            pagecontent_extension = None
+        try:
```

Comment:
I would write:

```python
        kwargs = {'app_label': self.model._meta.app_label, 'model_name': self.model.__name__.lower()}
        try:
            if pagecontent_extension:
                urls.append((page_content, admin_reverse(
                    '{app_label}_{model_name}_change'.format(**kwargs),
                    args=(pagecontent_extension.pk,)
                )))
            else:
                urls.append((page_content, '{0}?extended_object={1}'.format(
                    admin_reverse('{app_label}_{model_name}_add'.format(**kwargs)),
                    page_content.pk,
                )))
        except NoReverseMatch:  # pragma: no cover
            pass
```

this reduces code duplication and uses Python `.format()`-method instead of Py-2's `%`.
From Zen of Python: "Readability counts".

Reply:
Sorry for the late edit. Rechanged that snippet.